### ISSUE TITLE
Position offsets incorrect on analysis - not getting results from a phrase query using query_string

### ISSUE THREAD

**Issue by ontology-rory** [elastic/elasticsearch#31966]
**Elasticsearch version**: 6.2.1

**Plugins installed**: []

**JVM version** (`java -version`):
```
openjdk version "1.8.0_162"
OpenJDK Runtime Environment (Zulu 8.27.0.7-linux64) (build 1.8.0_162-b01)
OpenJDK 64-Bit Server VM (Zulu 8.27.0.7-linux64) (build 25.162-b01, mixed mode)
```

**OS version**: `Linux 3.10.0-514.10.2.el7.x86_64 #1 SMP Fri Mar 3 00:04:05 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux`

**Description of the problem including expected versus actual behavior**:

I have a document with the text `SAL_S8371 - SAL SUBTEND 0001` in it and perform a `query_string` search where the query is `"SAL_S8371 - SAL SUBTEND 0001"`. No documents are returned when I would expect to get a hit for this text.

When I run this text through the `_analyze` endpoint, I noticed that the position offsets are missing a  step when it encounters the dash, i.e. the position of '-' is 3 and the position of 'SAL' is 5 - there is no term at position 4.

Possibly related to #29021 ?

**Steps to reproduce**:

 1. create index
```
PUT testing
{
  "mappings": {
    "raw": {
      "_source": {
        "includes": [
          "*"
        ],
        "excludes": [
          "OntoAll"
        ]
      },
      "properties": {
        "OntoID": {
          "type": "keyword"
        },
        "OntoAll": {
          "type": "text"
        },
        "OntoFields": {
          "type": "nested",
          "properties": {
            "key": {
              "type": "keyword"
            },
            "value": {
              "type": "text",
              "copy_to": [
                "OntoAll"
              ]
            }
          }
        }
      }
    }
  },
  "settings": {
    "index": {
      "analysis": {
        "filter": {
          "OntoFilter": {
            "split_on_numerics": "true",
            "generate_word_parts": "true",
            "preserve_original": "true",
            "catenate_words": "false",
            "generate_number_parts": "true",
            "catenate_all": "false",
            "split_on_case_change": "true",
            "type": "word_delimiter_graph",
            "catenate_numbers": "false"
          }
        },
        "analyzer": {
          "default": {
            "filter": [
              "OntoFilter",
              "lowercase"
            ],
            "type": "custom",
            "tokenizer": "whitespace"
          }
        }
      }
    }
  }
}
```
 2. add document
```
PUT testing/raw/id1
{
  "OntoID": "S8371",
  "OntoFields": {
    "key": "prop",
    "value": "SAL_S8371 - SAL SUBTEND 0001"
  }
}
```
 3. run query and expected to get 1 result - but get 0
```
GET testing/_search
{
  "from": 0,
  "size": 10,
  "query": {
    "query_string": {
      "query": "\"SAL_S8371 - SAL SUBTEND 0001\"",
      "fields": [
        "OntoAll"
      ],
      "tie_breaker": 0,
      "default_operator": "and"
    }
  }
}
```
4. check the query via the `_analyze` endpoint - note position 4 is skipped
```
GET testing/_analyze
{
  "text": [
    "SAL_S8371 - SAL SUBTEND 0001"
  ]
}

{
  "tokens": [
    {
      "token": "sal_s8371",
      "start_offset": 0,
      "end_offset": 9,
      "type": "word",
      "position": 0,
      "positionLength": 3
    },
    {
      "token": "sal",
      "start_offset": 0,
      "end_offset": 3,
      "type": "word",
      "position": 0
    },
    {
      "token": "s",
      "start_offset": 4,
      "end_offset": 5,
      "type": "word",
      "position": 1
    },
    {
      "token": "8371",
      "start_offset": 5,
      "end_offset": 9,
      "type": "word",
      "position": 2
    },
    {
      "token": "-",
      "start_offset": 10,
      "end_offset": 11,
      "type": "word",
      "position": 3
    },
    {
      "token": "sal",
      "start_offset": 12,
      "end_offset": 15,
      "type": "word",
      "position": 5
    },
    {
      "token": "subtend",
      "start_offset": 16,
      "end_offset": 23,
      "type": "word",
      "position": 6
    },
    {
      "token": "0001",
      "start_offset": 24,
      "end_offset": 28,
      "type": "word",
      "position": 7
    }
  ]
}
```

I noticed that adding `"phrase_slop": 1` to the query then resulted in finding the document but feel this shouldn't be required - or that a greater slop maybe required for other queries where more position values are off.

Also added a simpler document with a similar query pattern and that works OK - i.e. get the expected result back

```
PUT testing/raw/id3
{
  "OntoID": "nice",
  "OntoFields": {
    "key": "prop",
    "value": "gogo - penguin"
  }
}
```

```
GET testing/_search
{
  "from": 0,
  "size": 10,
  "query": {
    "query_string": {
      "query": "\"gogo - penguin\"",
      "fields": [
        "OntoAll"
      ],
      "tie_breaker": 0,
      "default_operator": "and",
      "phrase_slop": 0
    }
  }
}
```

```
{
  "took": 67,
  "timed_out": false,
  "_shards": {
    "total": 5,
    "successful": 5,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": 1,
    "max_score": 0.8630463,
    "hits": [
      {
        "_index": "testing",
        "_type": "raw",
        "_id": "id3",
        "_score": 0.8630463,
        "_source": {
          "OntoID": "nice",
          "OntoFields": {
            "key": "prop",
            "value": "gogo - penguin"
          }
        }
      }
    ]
  }
}
```

**Comment by romseygeek** [comment]
Thanks for opening the bug.  This is lucene behaviour, and it definitely looks wrong.  I'll dig into it and open a JIRA issue.

**Comment by romseygeek** [comment]
I opened https://issues.apache.org/jira/browse/LUCENE-8395

**Comment by romseygeek** [comment]
This will be fixed in lucene 7.5
